### DISCUSSION TITLE
How do meta-transactions get from a relay onto a chain?

### DISCUSSION THREAD

**Discussion by GrahamMcBain** [WalletConnect/walletconnect-monorepo#1583] [Developer Support]
Hi,
I'm looking at the example relay server at https://github.com/WalletConnect/relay. I am expecting to see the relay connect to a node provider or GSN-like service (e.g. https://opengsn.org/) which would wrap the message in a transaction and pay for the gas.

Where is the relay "relaying" the messages?

**Comment by GrahamMcBain** [comment]
Related question https://github.com/orgs/WalletConnect/discussions/1579

**Comment by dbrans** [comment]
This was my question from discord. Thank you for posting it @GrahamMcBain. 

Hi, I'll elaborate more: I'm trying to understand how walletconnect v2 works end-to-end connecting a Dapp with a smart contract wallet (e.g. Argent). Here's my rough understanding with lots of gaps...

A user with a `smart contract wallet` (e.g. Argent) has an iphone with a `wallet client` (e.g. Argent mobile app) installed. The `wallet client` probably stores a private key corresponding to an eth-less EOA which is the owner of the `smart contract wallet`. 

The user visits a `Dapp` like https://app.ens.domains/. I imagine the interaction goes something like this:
1. User clicks "connect" and scans qr code with their `wallet client` (e.g. Argent mobile app).
1. `Dapp` and `wallet client `pair and establish a session (via a walletconnect `relay`??)
1. User clicks a button to perform an action like buying an ENS name.

At this point the `Dapp` needs the user's `smart contract wallet` to call some contract method (e.g. "ENSRegistry.register(string name) payable") with the required fee. However, a `Wallet Client` like Argent mobile does not have its own eth to pay for gas. So it cannot send a regular transaction like metamask. So:

4. `Dapp` asks `wallet client` to sign a meta-transaction (i.e., `address smartWallet, bytes data, nonce`) where the data encodes the call `0xENSRegistryAddress.register("coolname.eth")` with some `value`.
5. `wallet client` signs the meta-transaction with the private key corresponding to the owner of the `smart contract wallet`.
6. (`wallet client` or `dapp`)?? signs the passes the signed meta-transaction to a walletconnect `relay`??? which needs to get it on chain so:
7. `relay` passes meta-transaction to something like a `GSN service` (https://docs.opengsn.org/)???? which wraps the meta transaction in a transaction to the user's `smart contract wallet` and fronts the gas.
8. user's `smart contract wallet` executes the meta-transaction and refunds the `GSN forwarder` the gas.

I did not find anything describing a dapp-to-smart-contract-wallet flow in the walletconnect docs. Could you help me fill in these gaps? Any documentation, presentations, diagrams would be helpful.

**Reply by chadyj** [reply]
This might not exactly answer your question, but check out the specs section on our docs.  We go into lower level details about how WalletConnect works.

For example, this dives into the session proposal https://docs.walletconnect.com/2.0/specs/sign/session-proposal

**Reply by dbrans** [reply]
My question is less about how pairings and sessions get negotiated. It’s more about “How do transactions end up on chain and who pays for them?”

**Comment by dbrans** [comment]
Our fear is that “how transactions end up on chain” is proprietary as we cannot see the documentation or the code that does this.

**Comment by ganchoradkov** [comment]
Hey @GrahamMcBain @dbrans, to answer your questions, sign-client SDK does not communicate with the blockchain directly.
That is up to the Dapps & the wallets to do.
The Dapps prepare the transaction payload and send it via the sign-client to the paired wallet for approval.
When the transaction/request is approved the wallets can submit it to the blockchain.

We do have additional packages that provide blockchain connection such as our [universal-provider](https://github.com/WalletConnect/walletconnect-monorepo/tree/v2.0/providers/universal-provider) that establish JSON PRC connection to the blockchain node

**Reply by dbrans** [reply]
Thank you.
